ARCHITECTURE.md

```diff
@@ -291,3 +291,109 @@ Read more about [indexing].
+section of `Cargo.toml`
+
+## Target
+
+[CargoWorkspace.Target] is a thing that can be compiled to some binary artifact,
+e.g. executable binary, library, example library/executable, test executable, etc.
+Each target has crate root (see [CargoWorkspace.Target.crateRoot]). For example,
+`main.rs` or `lib.rs`.
+
+## Project model FAQ
+
+Q: What's the difference between [CargoProject] and [CargoWorkspace]?
+
+A: They are mostly the same, i.e. they are linked to each other. The most
+  difference is that [CargoWorkspace] is acquired from external tool,
+  and so may be `null` sometimes, while [CargoProject] is always persisted.
+
+Q: How does [CargoWorkspace] relate to `[workspace]` section in `Cargo.toml`?
+
+A: Each Cargo project has a workspace even if there is no `[workspace]` in
+  the `Cargo.toml`. With `[workspace]` the [CargoWorkspace] contains all
+  `[workspace.members]` packages.
+
+Q: What's the difference between [CargoWorkspace.Package] and [CargoWorkspace.Target]?
+
+A: [CargoWorkspace.Package] may contain multiple [CargoWorkspace.Target]s.
+  E.g. some package may contain common library target, multiple binary
+  targets that use it, test targets, benchmark targets, etc.
+
+Q: What is `Cargo.toml` dependency in the terms of this model?
+
+A: Dependency (that is `foo = "1.0"` in `Cargo.toml`) is a
+  [CargoWorkspace.Package] (of specified name and version) with one
+  _library_ target. See [CargoWorkspace.Package.dependencies]
+
+Q: What is a Rust crate in the terms of this model?
+
+A: It is always [CargoWorkspace.Target]. In the case of `extern crate foo;`
+  it is a library [CargoWorkspace.Target] (with a name `foo`) of some
+  dependency package of the current package.
+
+Q: What's the difference between [CargoWorkspace.Package.name] and [CargoWorkspace.Target.name]?
+
+A: [CargoWorkspace.Package.name] is a name of a dependency that should be mentioned in
+  `[dependencies]` section of `Cargo.toml`. [CargoWorkspace.Target.name] is a name
+  that visible in the Rust code, e.g. in `extern crate` syntax. Usually they are equal.
+  A name of a package can be specified by `[package.name]` property in `Cargo.toml`.
+  A name of a target can be specified in sections like `[lib]`, `[[bin]]`, etc.
+  Also, name of a dependency target can be changed.
+  Note that if a name of a target appears in the rust code, all `-` symbols are replaced with `_`.
+  To get such replaced names, use [CargoWorkspace.Target.normName] or [CargoWorkspace.Package.normName].
+  See [reference](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml)
+
+See Cargo's [workspace.rs](https://github.com/rust-lang/cargo/blob/d0f82841/src/cargo/core/workspace.rs).
+See Cargo's [package.rs](https://github.com/rust-lang/cargo/blob/d0f82841/src/cargo/core/package.rs).
+See Cargo's [manifest.rs](https://github.com/rust-lang/cargo/blob/d0f82841/src/cargo/core/manifest.rs#L228).
```

Comment:
Maybe it's better to reword it in the following way
`See Cargo's [workspace.rs], [package.rs] and [manifest.rs] for more details`
since we moved this doc from javadoc into ARCHITECTURE.md?